Mathlib/Data/Fin/Tuple/Sort.lean

```diff
@@ -107,6 +109,103 @@ theorem monotone_sort (f : Fin n → α) : Monotone (f ∘ sort f) := by
   exact (monotone_proj f).comp (graphEquiv₂ f).monotone
 #align tuple.monotone_sort Tuple.monotone_sort
 
+/-- All the elements `· ≤ a` appear the start of a sorted tuple -/
+theorem sort_lt_at_start_of_monotone {α} [LinearOrder α] (m : ℕ) (f : Fin m → α) (a : α)
+    (h_sorted : Monotone f)
+    (j : Fin m) (h : j < Fintype.card {i // f i ≤ a}) :
+    f j ≤ a := by
+  contrapose! h
+  rw [← Fin.card_Iio, Fintype.card_subtype]
+  refine Finset.card_mono (fun i => Function.mtr ?_)
+  simp_rw [Finset.mem_filter, Finset.mem_univ, true_and, Finset.mem_Iio]
+  exact fun hij => (h.trans_le <| h_sorted <| le_of_not_lt hij).not_le
+/- Proofs by Ruben Van de Velde, Eric {Rodriguez and Wieser}, Junyan Xu -/
+
+lemma Fintype.card_eq_subtypeSubtype_comm {α : Type} [Fintype α] (p q : α → Prop)
+    [DecidablePred p] [DecidablePred q] : Fintype.card (Subtype fun x => (p x) ∧ (q x)) =
+      Fintype.card (Subtype fun x => (q x) ∧ (p x)) := by
+  apply Fintype.card_congr
+  exact ⟨fun x => ⟨x, ⟨x.2.2, x.2.1⟩⟩, fun x => ⟨x, ⟨x.2.2, x.2.1⟩⟩,
+    fun _ => by dsimp, fun _ => by dsimp ⟩
+
+lemma Fintype.card_inter_eq_card_and {α : Type} [Fintype α] (p q : α → Prop)
+    [DecidablePred p] [DecidablePred q] : Fintype.card {i : (Subtype fun x => (p x)) // (q i)} =
+      Fintype.card (Subtype fun x => (q x) ∧ (p x)) := by
+  rw [← Fintype.card_eq_subtypeSubtype_comm p q]
+  refine' Fintype.card_eq.2 (Nonempty.intro ?_)
```

Comment:
```suggestion
  refine Fintype.card_congr ?_
```
(untested)

Note that generally you should prefer `refine` over `refine'`; the latter was for Lean 3 compatibility when we were porting.